sdk/nodejs/runtime/resource.ts

```diff
@@ -227,62 +233,139 @@ async function prepareResource(label: string, res: Resource, custom: boolean,
+    if (custom && opts.provider) {
+        const provider = opts.provider;
         const providerURN = await provider.urn.promise();
         const providerID = await provider.id.promise() || unknownValue;
         providerRef = `${providerURN}::${providerID}`;
     }
 
-    const dependencies: Set<URN> = new Set<URN>(explicitDependencies);
-    const propertyDeps: Record<string, URN[]> = {};
-    for (const key of Object.keys(propertyDependencies)) {
-        const urns = new Set<URN>();
-        for (const dep of propertyDependencies[key]) {
-            const urn = await dep.urn.promise();
-            urns.add(urn);
-            dependencies.add(urn);
-        }
-        propertyDeps[key] = Array.from(urns);
+    // Collect the URNs for explicit/implicit dependencies for the engine so that it can understand
+    // the dependency graph and optimize operations accordingly.
+
+    // The list of all dependencies (implicit or explicit).
+    const allDirectDependencies = new Set<Resource>(explicitDirectDependencies);
+
+    const allDirectDependencyURNs = await getResourceURNs(explicitDirectDependencies);
+    const propertyToDirectDependencyURNs = new Map<string, Set<URN>>();
+
+    for (const [propertyName, directDependencies] of propertyToDirectDependencies) {
+        addAll(allDirectDependencies, directDependencies);
+
+        const urns = await getResourceURNs(directDependencies);
+        addAll(allDirectDependencyURNs, urns);
+        propertyToDirectDependencyURNs.set(propertyName, urns);
     }
 
+    // Now await all dependencies transitively.  i.e. if one resource depends on some other
+    // component resource, then it will end up depending on all the children of that component
+    // resource as well.  That way, a parent acts as a "logical" collection of all those children
+    // and will not be considered complete until all of the children are complete as well.
+    await getResourceURNs(getTransitivelyDependentResources(allDirectDependencies, res));
+
     return {
         resolveURN: resolveURN!,
         resolveID: resolveID,
         resolvers: resolvers,
         serializedProps: serializedProps,
         parentURN: parentURN,
         providerRef: providerRef,
-        dependencies: dependencies,
-        propertyDependencies: propertyDeps,
+        allDirectDependencyURNs: allDirectDependencyURNs,
+        propertyToDirectDependencyURNs: propertyToDirectDependencyURNs,
     };
 }
 
+function addAll<T>(to: Set<T>, from: Set<T>) {
+    for (const val of from) {
+        to.add(val);
+    }
+}
+
+async function getResourceURNs(resources: Set<Resource>) {
+    const result = new Set<URN>();
+    for (const resource of resources) {
+        result.add(await resource.urn.promise());
+    }
+
+    return result;
+}
+
+function getTransitivelyDependentResources(resources: Set<Resource>, thisResource: Resource) {
+    // keep iterating until we reach a fixed point.
```

Comment:
It's kind of a small thing--I don't want us to optimize where we don't need to--but this seems unnecessarily expensive. Couldn't we just do iterative DFS using a stack? Even the recursive implementation seems preferable.